### PR TITLE
Never re-use task workers for actors or GPU tasks

### LINKED ISSUE #20554: [Bug] OwnerDiedError when a task worker is reused as an actor
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

This bug happens when a worker creates an ObjectRef while executing a task, then later gets reused as an actor. If the actor goes out of scope, Ray force-kills the process without waiting for any owned ObjectRefs to drain. Later, this will trigger an `OwnerDiedError` if another process tries to get the ObjectRef because the object's owner has been killed.

### Versions / Dependencies

Ray 2.0dev

### Reproduction script

```python
import ray
import numpy as np


@ray.remote
def create_ref():
    ref = ray.put(np.zeros(100_000_000))
    return ref

@ray.remote
class Actor:
    def __init__(self):
        return

    def foo(self):
        return

ray.init(num_cpus=1, object_store_memory=1000_000_000)  # 1 CPU so we only start 1 worker.
wrapped_ref = create_ref.remote()
print(ray.get(ray.get(wrapped_ref)))

# create_ref worker gets reused as an actor.
a = Actor.remote()
ray.get(a.foo.remote())
# Actor will get force-killed.
del a

# Flush the object store.
for _ in range(10):
    ray.put(np.zeros(100_000_000))

# Object has been evicted and owner has died. Throws OwnerDiedError.
print(ray.get(ray.get(wrapped_ref)))
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### LINKED ISSUE #22504: [Bug] torch.cuda.is_available() is False after running CPU task
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

When using PyTorch on Ray, mixing CPU and GPU tasks / actors leads to the GPU tasks being unable to properly initialize CUDA. 

Running a CPU ray task that uses PyTorch followed by a GPU task or actor leads to the latter being unable to init CUDA.

What should happen: regardless of whether any CPU tasks were run before the GPU task, the GPU task / actor should be able to properly init CUDA.

### Versions / Dependencies

```
ray==1.10.0
torch==1.10.2+cu111
```

### Reproduction script

```
import torch
import ray

ray.init("auto")

@ray.remote
def test_fn():
    print(torch.cuda.is_available())

@ray.remote(num_gpus=1)
def test_gpu_fn():
    print(torch.cuda.is_available())

ray.get(test_fn.remote())
ray.get(test_gpu_fn.remote())
```

prints: `False False` (expected: `False True`)

Also happens if the GPU task is changed to an actor:

```
import torch
import ray

ray.init("auto")

@ray.remote
def test_fn():
    print(torch.cuda.is_available())

@ray.remote(num_gpus=1)
class GpuActor(object):
    def test_gpu_fn(self):
        print(torch.cuda.is_available())

ray.get(test_fn.remote())

actor = GpuActor.remote()
ray.get(actor.test_gpu_fn.remote())
```

prints: `False False`

However, if we change the CPU task to an actor, it works as expected:

```
import torch
import ray

ray.init("auto")

@ray.remote
class CpuActor(object):
    def test_fn(self):
        print(torch.cuda.is_available())

@ray.remote(num_gpus=1)
class GpuActor(object):
    def test_gpu_fn(self):
        print(torch.cuda.is_available())

cpu_actor = CpuActor.remote()
ray.get(cpu_actor.test_fn.remote())

actor = GpuActor.remote()
ray.get(actor.test_gpu_fn.remote())
```

prints: `False True`

Furthermore, if we run the GPU task first, it works correctly:

```
import torch
import ray

ray.init("auto")

@ray.remote
def test_fn():
    print(torch.cuda.is_available())

@ray.remote(num_gpus=1)
def test_gpu_fn():
    print(torch.cuda.is_available())

ray.get(test_gpu_fn.remote())
ray.get(test_fn.remote())
```

prints: `True False`

### Anything else

This is running in Docker on AWS (g4dn.8xlarge)

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Don't re-use task workers for actors, since those workers may own objects that will be lost on actor exit.

This adds a slight performance penalty for actor startup.

## Related issue number

Closes https://github.com/ray-project/ray/issues/20554
Closes https://github.com/ray-project/ray/issues/22504

### CODE REVIEW

**Comment by ericl** [comment]
@iycheng , how do you trigger the actor performance tests? I'd like to check that this doesn't regress the performance of starting a large number of actors.

**Comment by iycheng** [comment]
@ericl  you can run this test https://buildkite.com/ray-project/periodic-ci/builds/2809#be9c211b-459e-4ed7-a5ed-b7675b23a813
But I think it might just be the same since this test is launching a lot of actors from an empty cluster. So no reusing workers should happen there.

**Comment by ericl** [comment]
Thanks. Do we have documentation on how to trigger it? I think it might be relevant because prestarted workers will no longer be used for actors. Not sure if they were in the first place though.

**Comment by iycheng** [comment]
I can't find it now. But I remember @rkooo567  has doc about that.

File: python/ray/tests/test_actor.py [hunk 1]
```diff
@@ -55,6 +55,39 @@ def get_val(self):
...
+    ray.get(a.foo.remote())
+    # Actor will get force-killed.
+    del a
+
+    # Flush the object store.
+    for _ in range(10):
+        ray.put(np.zeros(100_000_000))
+
+    # Object has been evicted and owner has died. Throws OwnerDiedError.
+    print(ray.get(ray.get(wrapped_ref)))
```
Comment (stephanie-wang) [inline python/ray/tests/test_actor.py:88]: Nice test!

File: src/ray/common/task/task_spec.h [hunk 2]
```diff
@@ -341,8 +341,10 @@ class WorkerCacheKey {
   ///
   /// worker. \param serialized_runtime_env The JSON-serialized runtime env for this
   /// worker. \param required_resources The required resouce.
+  /// worker. \param is_actor Whether the worker will be an actor.
```
Comment (stephanie-wang) [inline src/ray/common/task/task_spec.h:344]: Add a comment to explain why we hash on this?

**Review by stephanie-wang — APPROVED** [review]

File: python/ray/tests/test_actor.py (hunk 1 above)
Comment (ericl) [reply python/ray/tests/test_actor.py:88]: Copy pasted from your issue haha.

File: python/ray/tests/test_metrics.py [hunk 3]
```diff
@@ -84,7 +84,7 @@ def f(self):
     for stats in reply.core_workers_stats:
         if stats.webui_display[""] == '{"message": "test", "dtype": "text"}':
             target_worker_present = True
-            assert stats.pid == worker_pid
+        #            assert stats.pid == worker_pid
```
Comment (jjyao) [inline python/ray/tests/test_metrics.py:87]: ? should we just remove the code?

**Review by clarkzinzow — APPROVED** [review]

**Review by jjyao — APPROVED** [review]

**Comment by ericl** [comment]
Hmm bumping Tune size of BOHB example just in case, seems it timed out.

**Pull request merged**